scipy/stats/_sampling.py

```diff
@@ -0,0 +1,1234 @@
+           "Random variate  generation by numerical inversion when only the
+           density is known." ACM Transactions on Modeling and Computer
+           Simulation (TOMACS) 20.4 (2010): 1-25.
+    .. [2] Hörmann, Wolfgang, Josef Leydold and Gerhard Derflinger.
+           "Automatic nonuniform random number generation."
+           Springer, 2004.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>> from scipy.stats.sampling import FastGeneratorInversion
+
+    Let's start with a simple example to illustrate the main features:
+
+    >>> gamma_dist = FastGeneratorInversion('gamma', (1.5, ))
+    >>> r = gamma_dist.rvs(size=1000)
+
+    The mean should be approximately equal to the shape parameter 1.5:
+
+    >>> r.mean()
+    1.52423591130436  # may vary
+
+    Similarly, we can draw a sample based on quasi-random numbers:
+
+    >>> r = gamma_dist.qrvs(size=1000)
+    >>> r.mean()
+    1.4996639255942914  # may vary
+
+    Compare the PPF against approximation ppf_fast.
+
+    >>> q = [0.001, 0.2, 0.5, 0.8, 0.999]
+    >>> np.max(np.abs(gamma_dist.ppf(q) - gamma_dist.ppf_fast(q)))
+    4.313394796895409e-08
+
+    To confirm that the numerical inversion is accurate, we evaluate the
+    approximation error (u-error), which should be below 1e-10 (for more
+    details, refer to the documentation of `evaluate_error`):
+
+    >>> gamma_dist.evaluate_error()
+    (7.446320551265581e-11, nan)  # may vary
+
+    Note that the location and scale can be changed without instantiating a
+    new generator:
+
+    >>> gamma_dist.loc = 2
+    >>> gamma_dist.scale = 3
+    >>> r = gamma_dist.rvs(size=1000)
+
+    The mean should be approximately 2 + 3*1.5 = 6.5.
+
+    >>> r.mean()
+    6.399549295242894  # may vary
+
+    Let us also illustrate how truncation can be applied:
+
+    >>> trunc_norm = FastGeneratorInversion('norm', domain=(3, 4))
+    >>> r = trunc_norm.rvs(size=1000)
+    >>> 3 < r.min() < r.max() < 4
+    True
+
+    Check the mean:
+
+    >>> r.mean()
+    3.250433367078603  # may vary
+
+    >>> stats.norm.expect(lb=3, ub=4, conditional=True)
+    3.260454285589997
+
+    Compare `ppf` and `ppf_fast`:
+
+    >>> trunc_norm.ppf(0.5)
+    3.1984741317619285
+    >>> trunc_norm.ppf_fast(0.5)
+    3.198474131740042
+
+    In this particular, case, `scipy.stats.truncnorm` could also be used to
+    generate truncated normal random variates. We can verify that
+    the CDFs coincide up to a small numerical error:
+
+    >>> trunc_norm.cdf(3.5) - stats.truncnorm(3, 4).cdf(3.5)
+    -1.0325074129013956e-14
+
+    """
+
+    def __init__(
+        self,
+        distname,
+        args=(),
+        *,
+        domain=None,
+        loc=0,
+        scale=1,
+        ignore_shape_range=False,
+        random_state=None,
+    ):
+
+        if not isinstance(distname, str):
+            raise TypeError("distname must be a string.")
+        if distname not in set(_distn_names):
```

Comment:
Why not validate this immediately against `PINV_CONFIG`? That's now done further down, which is (a) double validation, and (b) may not be fully robust for the intermediate code.